### DISCUSSION TITLE
How to use autocomplete API in an array

### DISCUSSION THREAD

**Discussion by limegorilla** [directus/directus#10128] [Archive]
Anyone have any ideas on the Autocomplete API?
I'm trying to get this API: https://huxley2.azurewebsites.net/crs/

Returns this (if input is 'london')
```
[
    {
        "stationName": "London Blackfriars",
        "crsCode": "BFR"
    },
    {
        "stationName": "London Bridge",
        "crsCode": "LBG"
    },
    {
        "stationName": "London Cannon Street",
        "crsCode": "CST"
    },
    {
        "stationName": "London Charing Cross",
        "crsCode": "CHX"
    },
    {
        "stationName": "London Euston",
        "crsCode": "EUS"
    },
    {
        "stationName": "London Fenchurch Street",
        "crsCode": "FST"
    },
...
]
```

How I have it setup just currently isn't working - see the image.

![Screenshot 2021-11-29 at 11 09 40 am](https://user-images.githubusercontent.com/29333568/143857568-4d8b241a-9e5f-44db-8f7e-7f73a3b772bc.png)

In a perfect world, I would search using the `stationName` and then save `crsCode` to the Database, but if I can just get the stationName that's fine (I can probably translate this to something else later)

**Comment by rijkvanzanten** [comment]
Can you try setting the value path to a specific index? For example `[0].crsCode`

**Reply by limegorilla** [reply]
Tried this - nothing is coming up under the text box still when I type :/

**Reply by limegorilla** [reply]
This is currently how I have it setup ![Screenshot 2021-11-29 at 4 19 03 pm](https://user-images.githubusercontent.com/29333568/143904034-78e86f04-9844-4131-a232-f9612ff43866.png)

**Reply by azrikahar** [reply]
@limegorilla your URL format is correct, but I believe you don't need to prefix `[]` here. To understand what is the "shape" of the payload (which is actually what you've already seen when calling the API directly), you can try leaving both `Results Path` and `Value Path` blank, where you should get the following result:

![image](https://user-images.githubusercontent.com/42867097/143971110-c7977894-b614-4a40-97e7-32ff086762d4.png)

To clarify, leaving results path meant getting the whole thing as the result, which is what you want here since the API doesn't return the array in something like the following, where you will need to set `items` as results path:

```js
{
  "items": [
    {"title": "Item 1"},
    {"title": "Item 2"}
   }
}
```

As for value path, it is from per-item perspective. Since for each item in your current API returns `stationName` and `crsCode`, the latter should be what you should key in for Value Path, without any `[]` prefix.

Final result with the config and how it'll look:

![chrome_h5v7aAd8M7](https://user-images.githubusercontent.com/42867097/143972291-9dd91f96-573f-49fe-9e62-a378810ef913.png)

![OFzI9HgSyo](https://user-images.githubusercontent.com/42867097/143972302-7c62d616-0435-4121-bcfc-acd9d35a1960.gif)

**Reply by limegorilla** [reply]
Brilliant this is perfect!

Really should have tried it without anything in there - sorry about that!

**Reply by rickysullivan** [reply]
What if you wanted to store the value, in my case it's an integer (id) and keep displaying the string value (a).

![image](https://user-images.githubusercontent.com/437480/207824574-5e3e1655-0005-45e5-8741-6763b7047256.png)

When a selection is made, currently it shows this:

![image](https://user-images.githubusercontent.com/437480/207824690-511ea70c-776d-49b8-b139-fda9e14038a2.png)

I'd love to be able to store my result as `{ "a": "Some fancy address", "id": 2272567 }` and using a display template:

![image](https://user-images.githubusercontent.com/437480/207832493-5a40e1df-ce66-4044-b818-7c653f5cc89b.png)

Render:

![image](https://user-images.githubusercontent.com/437480/207834791-4c38e362-530f-4bae-a22e-014aaa070320.png)

**Reply by quentinglorieux** [reply]
did you manage to do it rick ?

**Reply by rickysullivan** [reply]
Sort of.

Short story long:

I have cloned a lot of the code from the autocomplete interface, creating a custom one. I have also set my field/column to `json`. Which needs a default value of `{ "id": "", "a": ""}` else the field disappears from the UI.

<img width="932" alt="image" src="https://user-images.githubusercontent.com/437480/211635651-4fbd413c-6c79-4b0d-836a-f76d29a8f523.png">

I change the `v-list` to emit the entire result:

```js
<v-list v-if="results.length > 0">
  <v-list-item v-for="result of results" :key="result.value" @click="() => emitValue(result)">
    <v-list-item-content>{{ textPath ? result.text : result.value }}</v-list-item-content>
  </v-list-item>
</v-list>
```

Then for the `v-input`, the `model-value` is either the `value.text` when creating/searching or `value` if editing an existing user:

```js
<v-input
    :placeholder="placeholder"
    :disabled="disabled"
    :class="font"
    :model-value="value.text ?? value "
    :dir="direction"
    @update:model-value="onInput"
    @focus="activate"
>
  <template v-if="iconLeft" #prepend>
    <v-icon :name="iconLeft"/>
  </template>
  <template v-if="iconRight" #append>
    <v-icon :name="iconRight"/>
  </template>
</v-input>
```

What is annoying is that I've tried to mark the field as being required, but I can still submit with no value, I presume this is because o the default value having been set.

I then format the JSON for display:

<img width="919" alt="image" src="https://user-images.githubusercontent.com/437480/211638844-5da12924-843a-4ccc-b321-63127d43d714.png">

Which works well.

<img width="922" alt="image" src="https://user-images.githubusercontent.com/437480/211638743-3366d82e-cf63-4b03-a558-31d732331e0c.png">

The reason I needed to keep both values ("a" and "id") is that I've got a Workflow that runs when a new Customer
 is added to the collection. 

During this workflow, I perform an address lookup that grabs extra details about the customer's address (long, lat):

<img width="1406" alt="image" src="https://user-images.githubusercontent.com/437480/211640304-1a4266bb-9cc6-43b9-9bea-9d73b3eb2e55.png">

This allows me to then have a custom map display for all the Addresses in my system:

<img width="1413" alt="image" src="https://user-images.githubusercontent.com/437480/211640621-1ce15579-74e0-46d0-9ec6-daa76a43366c.png">

And also helps to split the address into individual columns/fields:

<img width="948" alt="image" src="https://user-images.githubusercontent.com/437480/211641017-07128b5b-1949-4087-bbf9-c4dbca6cdc46.png">

**Reply by rickysullivan** [reply]
Here's a link to my custom interface: https://github.com/rickysullivan/autocomplete-alt

**Reply by jeffersonsolorzano** [reply]
@rickysullivan I can see that you have a lat-long field did you create another field for the map interface and inserted those values there too, or are you getting the lat-long values form those fields and use it for displaying the point in the map?

**Reply by rickysullivan** [reply]
I have another field for the map which uses this lat long values to display a pin.

**Reply by jeffersonsolorzano** [reply]
How did you do that?
If i change the value in the map, does it change in the field too?

**Reply by rickysullivan** [reply]
Workflows. I’ll have to fire up my other machine and screen shot it.

**Reply by jeffersonsolorzano** [reply]
Sure, I'm looking for a way to change the values of the fields in realtime and fill all the fields information with the captured lat and long values, i could get all this info with this endpoint:
```
https://nominatim.openstreetmap.org?addressdetails=1&extratags=1&namedetails=1&q=16.715456736311854, -92.71498695945886&format=json&limit=1
```
 and it will give me all this info:
<img width="1023" alt="image" src="https://github.com/directus/directus/assets/65744166/c255b179-bd26-4db2-9845-d66d29abe042">

**Comment by paesjr** [comment]
Hi @rickysullivan, what API you are using? I am also in NZ (New Plymouth) and I need to implement a similar solution. 
I'm a new developer and also new to Directus by the way.

**Reply by rickysullivan** [reply]
I was using the https://addressfinder.nz/features/address-verification/ service. There’s also Addy Solutions: https://www.addysolutions.com/

**Comment by paesjr** [comment]
Thanks for the fast reply, Ricky! 😊

From: Ricky Sullivan Himself ***@***.***>
Date: Friday, 10 November 2023 at 3:08 PM
To: directus/directus ***@***.***>
Cc: Ricardo Da Motta Paes Jr ***@***.***>, Comment ***@***.***>
Subject: Re: [directus/directus] How to use autocomplete API in an array (Discussion #10128)

I was using the https://addressfinder.nz/features/address-verification/ service. There’s also Addy Solutions: https://www.addysolutions.com/

**Reply by rickysullivan** [reply]
No worries.
